File: 0677_Map Sum Pairs/MapSumPaires_Trie.cs
public class MapSum {
    private Trie root = new Trie(0);
    private Dictionary<string,int> dict = new Dictionary<string,int>();
    
    /** Initialize your data structure here. */
    public MapSum() {
        
    }
    
    public void Insert(string key, int val) {
        
        var offset = val;
        if(dict.ContainsKey(key)){
            offset -= dict[key];
            dict[key] = val;
        }else{
            dict.Add(key, val);
        }
        
        
        var p = root;
        foreach(var c in key)
        {
            if(p.Next[c - 'a'] == null) 
                p.Next[c - 'a'] = new Trie(offset);
            else
                p.Next[c - 'a'].Value += offset;
            
            p = p.Next[c - 'a'];
        }
    }
    
    public int Sum(string prefix) {
        
        var p = root;
        int sum = 0;
        foreach(var c in prefix)
        {
            if(p.Next[c - 'a'] != null){
                p = p.Next[c - 'a'];
                sum = p.Value;
            }else{
                sum = 0;
                break;
            }
        }
        
        return sum;
    }
}

public class Trie {
    public Trie(int v){
        this.Value = v;
    }
    public Trie[] Next = new Trie[26];
    public int Value = 0;
}

/**
 * Your MapSum object will be instantiated and called as such:
 * MapSum obj = new MapSum();
 * obj.Insert(key,val);
 * int param_2 = obj.Sum(prefix);
 */

File: 0677_Map Sum Pairs/MapSumPaires_Trie.java
class MapSum {
    private TireNode root;
    private Map<String,Integer> map;
    public MapSum() {
        root = new TireNode(0);
        map = new HashMap<>();
    }
    
    public void insert(String key, int val) {
        int offset = val - map.getOrDefault(key, 0);
        map.put(key, val);
        
        TireNode p = root;
        
        for(char c : key.toCharArray()) {
            if(p.children[c-'a'] == null) 
                p.children[c-'a'] = new TireNode(offset);
            else {
                p.children[c-'a'].val += offset;
            }
            p = p.children[c-'a'];
        }
    }
    
	public int sum(String prefix) {
		TireNode p = root;
		int sum = 0;
		for (char c : prefix.toCharArray()) {
			if (p.children[c - 'a'] != null) {
				p = p.children[c - 'a'];
                sum = p.val;
			} else {
				return 0;
			}
		}

		return sum;
	}
    
	private class TireNode {
	    TireNode[] children = new TireNode[26];
	    int val;
	    public TireNode(int v) {
	        val = v;
	    }
	}
}